examples/tutorials/nb_python/ECCV_2020_Interactivity.py

```diff
@@ -0,0 +1,2093 @@
+# @markdown (double click to show code)
+
+# Event handler for dropdowns displaying file-based object handles
+def on_file_obj_ddl_change(ddl_values):
+    global sel_file_obj_handle
+    sel_file_obj_handle = ddl_values["new"]
+    return sel_file_obj_handle
+
+
+# Event handler for dropdowns displaying prim-based object handles
+def on_prim_obj_ddl_change(ddl_values):
+    global sel_prim_obj_handle
+    sel_prim_obj_handle = ddl_values["new"]
+    return sel_prim_obj_handle
+
+
+# Event handler for dropdowns displaying asset handles
+def on_prim_ddl_change(ddl_values):
+    global sel_asset_handle
+    sel_asset_handle = ddl_values["new"]
+    return sel_asset_handle
+
+
+# Build a dropdown list holding obj_handles and set its event handler
+def set_handle_ddl_widget(obj_handles, handle_types, sel_handle, on_change):
+    sel_handle = obj_handles[0]
+    descStr = handle_types + " Template Handles:"
+    style = {"description_width": "300px"}
+    obj_ddl = widgets.Dropdown(
+        options=obj_handles,
+        value=sel_handle,
+        description=descStr,
+        style=style,
+        disabled=False,
+        layout={"width": "max-content"},
+    )
+
+    obj_ddl.observe(on_change, names="value")
+    return obj_ddl, sel_handle
+
+
+def set_button_launcher(desc):
+    button = widgets.Button(description=desc, layout={"width": "max-content"},)
+    return button
+
+
+def make_sim_and_vid_button(prefix, dt=1.0):
+    if not HAS_WIDGETS:
+        return
+
+    def on_sim_click(b):
+        observations = simulate(sim, dt=dt)
+        make_video_cv2(observations, prefix=prefix, open_vid=True, multi_obs=False)
+
+    sim_and_vid_btn = set_button_launcher("Simulate and Make Video")
+    sim_and_vid_btn.on_click(on_sim_click)
+    ipydisplay(sim_and_vid_btn)
+
+
+def make_clear_all_objects_button():
+    if not HAS_WIDGETS:
+        return
+
+    def on_clear_click(b):
+        remove_all_objects(sim)
+
+    clear_objs_button = set_button_launcher("Clear all objects")
+    clear_objs_button.on_click(on_clear_click)
+    ipydisplay(clear_objs_button)
+
+
+# Builds widget-based UI components
+def build_widget_ui(obj_attr_mgr, prim_attr_mgr):
+    # Holds the user's desired file-based object template handle
+    global sel_file_obj_handle
+    sel_file_obj_handle = ""
+
+    # Holds the user's desired primitive-based object template handle
+    global sel_prim_obj_handle
+    sel_prim_obj_handle = ""
+
+    # Holds the user's desired primitive asset template handle
+    global sel_asset_handle
+    sel_asset_handle = ""
+
+    # Construct DDLs and assign event handlers
+    # All file-based object template handles
+    file_obj_handles = obj_attr_mgr.get_file_template_handles()
+    # All primitive asset-based object template handles
+    prim_obj_handles = obj_attr_mgr.get_synth_template_handles()
+    # All primitive asset handles template handles
+    prim_asset_handles = prim_attr_mgr.get_template_handles()
+    if not HAS_WIDGETS:
```

Comment:
The defaults I picked earlier.

Reply:
If we factor this out, I'll need to put another shim conditional under all calls, i guess? Or make a shim function